python/sbt-vim.py

```diff
@@ -95,19 +95,23 @@ def close(self):
 
   def command(self, cmd):
     to = self.proc.stdin
+    send = lambda s: to.write(s.encode('utf-8'))
     fr = self.proc.stdout
+    recv = lambda : fr.readline().decode('utf-8')
+
     marker = str(uuid.uuid4())
-    to.write(cmd)
-    to.write("\n")
-    to.write("eval print(\"%s\\n\")\n" % marker)
-    fr.readline()
-    line = fr.readline()
-    while line:
+
+    send(cmd + '\n')
+    send('eval print("{:s}\\n")\n'.format(marker))
+    to.flush()
+
+    while True:
+      line = recv()
+      if not line:
+        continue
```

Comment:
What's this for?

Reply:
This patch is needed because Python3 treats subprocess pipes as *byte* streams by default, rather than character streams in Python2, so all out-going and in-coming messages need to be converted between unicode strings and byte-sequences. That translation was easier to encapsulate within the new send() and recv() lambda-functions than repeat across multiple write()/read() calls. Python3 will also not automatically flush the output stream, hence the explicit call to "flush()", without which vim will simply hang.

Reply:
Sorry, I was only asking about the “if not line: continue”. That doesn’t seem to have any analog in the old code and would cause the loop to hang if for some reason the marker never shows up, right?

Reply:
Thanks for clarifying - sorry I didn't understanding the first time.
I'm not sure either version of the code is perfectly robust - each makes assumptions about the sbt subprocess not hanging mid-way through its output stream. I think that the situations where sbt can't be relied upon to print the uuid-marker are equally likely to cause readline() to get stuck before it receives an end-of-line. What I prefer about the patched version of this loop is that it doesn't call self.proc.stdout.readline() from multiple places. I expect if the sbt subprocess terminates, both versions of the code would end up with an exception being thrown.

Reply:
OK